<a href="https://colab.research.google.com/github/Kekichka/AI/blob/labs/Basic_AI_practise_1_2025_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Оцінювання даної роботи буде ураховувати якість Вашого коду та текстових відповідей. При наявності ЛИШЕ коду , робота оцінюватися НЕ БУДЕ.**  

**Оформлення текстових відповідей повинно проводитися в текстових комірках ноутбука з використанням синтаксису [LaTeX](https://en.wikibooks.org/wiki/LaTeX/Basics) для оформлення тексту та формул.**

**Роботи ідентифіковані як ІДЕНТИЧНІ одразу отримують 0 балів.**

**При реалізації завдань потрібно використовувати методи з бібліотеки numpy та pandas.**

**Рішення з циклами будуть оцінені в 0 балів.**

# I. NumPy.

0. Імпортувати numpy загальноприйнятим способом.

In [3]:
import numpy as np

1. Визначити список з оцінками студента(ки) за семестр. Використовуючи його, створити одновимірний масив numpy.

In [ ]:
mygrades = [5, 4, 4, 5, 5]
mygrades_array = np.array(mygrades)
print("Hopefully my grades will be like this: " + str(mygrades_array))

Hopefully my grades will be like this: [5 4 4 5 5]


2. Визначити список зі списками, де кожен рядок списку формально означає магазин, а кожна колонка - певний продукт (батон, яйця першої категорії, молоко та ковбаса), а елементами на пересічі є ціни на ці продукти відповідно. Використовуючи цей список, створити матрицю (двовимірний масив) numpy.

In [ ]:
prices = [
    [15.5, 30.0, 25.0, 120.0],
    [16.0, 28.5, 26.0, 115.0],
    [14.8, 29.0, 24.5, 122.0],
]
prices_matrix = np.array(prices)
print(prices)

[[15.5, 30.0, 25.0, 120.0], [16.0, 28.5, 26.0, 115.0], [14.8, 29.0, 24.5, 122.0]]


3. Отримайте тип даних значень з масивів завдань 2 та 3. Виведіть їх на екран.

In [ ]:
print(mygrades_array.dtype)
print(prices_matrix.dtype)

int64
float64


4. Отримайте форми (кортеж з кількістю рядків та колонок) масивів завдань 2 та 3. Виведіть їх на екран.

In [ ]:
print(mygrades_array.shape)
print(prices_matrix.shape)

(5,)
(3, 4)


5. Створіть одновимірний масив за допомогою спеціальної функції numpy, який би відображав динаміку стабільно зростаючого (з рівними проміжками) прибутку з продажів  за тиждень, де у перший день не було продажів, а в останній день тижня вдалося заробити 1000 грн 50 коп.

In [ ]:
weekly_sales = np.linspace(0, 1000.50, 7)
print(weekly_sales)

[   0.    166.75  333.5   500.25  667.    833.75 1000.5 ]


6. Створіть два масиви. Використовуючи їх, продемонструйте відмінність вертикального та горизонтального обʼєднання масивів.

In [ ]:
cheeseburgers = [7, 8, 3, 5]
hamburgers = [1, 4, 9, 6]
vertical = np.vstack((cheeseburgers, hamburgers))
horisontal = np.hstack((cheeseburgers, hamburgers))
print(vertical)
print(horisontal)

[[7 8 3 5]
 [1 4 9 6]]
[7 8 3 5 1 4 9 6]


7. Визначити функцію, яка приймає на вході масив і транспонує його. Скористайтесь методом reshape. І не забудьте почати з docstrings.

In [ ]:
def transpose_matrix(matrix):
    """
    Transposes a given 2D numpy array (matrix).

    Parameters:
        matrix (numpy.ndarray): A 2D array to be transposed.

    Returns:
        numpy.ndarray: The transposed matrix.
    """
    return matrix.T

hours_worked = np.array([
    [3, 4, 3],
    [4, 2, 5],
    [1, 3, 1],
    [4, 2, 1]
])

print("Original matrix (hours worked over days on subjects):")
print(hours_worked)

transposed_hours = transpose_matrix(hours_worked)

print("\nTransposed matrix (hours worked per day on each subject):")
print(transposed_hours)

Original matrix (hours worked over days on subjects):
[[3 4 3]
 [4 2 5]
 [1 3 1]
 [4 2 1]]

Transposed matrix (hours worked per day on each subject):
[[3 4 1 4]
 [4 2 3 2]
 [3 5 1 1]]


8. Створіть два масиви. Використовучи ці масиви, продемонструйте 1) операцію по-елементного додавання, 2) операцію по-елементного віднімання, 3) множення масиву на число, 4) операцію по-елементного множення, і 5) матричного множення.

In [16]:
shopping_cart_makeup = [6, 5, 7, 8, 1]
shopping_cart_food = [2, 4, 9, 3, 5]

shopping_cart_makeup = np.array(shopping_cart_makeup)
shopping_cart_food = np.array(shopping_cart_food)

element_add = shopping_cart_makeup + shopping_cart_food
element_subtract = shopping_cart_makeup - shopping_cart_food
multiply = shopping_cart_food * 7
element_multiply = shopping_cart_makeup * shopping_cart_food
matrix_multiply = np.dot(shopping_cart_makeup, shopping_cart_food)

print("Makeup and food items :", element_add)
print("Makeup without food :", element_subtract)
print("Food for seven people:", multiply)
print("Multiplication :", element_multiply)
print("Matrix multiplication:", matrix_multiply)


Makeup and food items : [ 8  9 16 11  6]
Makeup without food : [ 4  1 -2  5 -4]
Food for seven people: [14 28 63 21 35]
Multiplication : [12 20 63 24  5]
Matrix multiplication: 124


9. Створіть двовимірний масив (матрицю) та розрахуйте: 1) мінімальне число, 2) максимальне число, 3) суму чисел, 4) мінімальні числа для кожного рядка, 5) максимальні числа для кожного стовпчика.

In [17]:
cheeseburgers = np.array([
    [12, 15, 7, 9],
    [8, 14, 10, 5],
    [19, 3, 11, 6],
    [13, 18, 4, 2]
])

min = np.min(cheeseburgers)
max = np.max(cheeseburgers)
sum = np.sum(cheeseburgers)
min_row = np.min(cheeseburgers, axis=1)
max_column = np.max(cheeseburgers, axis=0)

print("Min cheeseburgers eaten per day: ", min)
print("Max cheeseburger eaten per day: ", max)
print("Cheeseburgers eaten: ", sum)
print("Min cheeseburgers for row: ", min_row)
print("Max cheeseburgers for column: ", max_column)

Min cheeseburgers eaten per day:  2
Max cheeseburger eaten per day:  19
Cheeseburgers eaten:  156
Min cheeseburgers for row:  [7 5 3 2]
Max cheeseburgers for column:  [19 18 11  9]


10. Створіть матрицю , де деякі елементи повторюються. Знайдіть унікальні значення елементів цієї матриці та їхні частоти.

In [18]:
movies = np.array([
    [1, 2, 3, 1],
    [2, 1, 3, 2],
    [3, 1, 2, 3],
    [2, 3, 1, 2]
])

unique_values, counts = np.unique(movies, return_counts=True)

for value, count in zip(unique_values, counts):
    print(f"Movie: {value}, times watched: {count}")

Movie: 1, times watched: 5
Movie: 2, times watched: 6
Movie: 3, times watched: 5


# II. Importing. Creating DF

0. Імпортувати бібліотеку pandas

In [2]:
import pandas as pd

11. Створити тестовий датафрейм зі списку, що містить записи пацієнтів про їхні прізвища, імена, дату народження, масу тіла та наявність медичного страхування. Одне зі значень зробить пустим. Ознайомтесь з типом кожної колонки.

In [ ]:
patients_info = {
    'Last Name': ['Cookie', 'Mango', 'Sherbet', 'Shining', 'Hollyberry'],
    'First Name': ['Cadence', 'Candy', 'Pookie', 'Jack', 'Meow'],
    'Birth Date': ['1977-07-07', '1990-08-21',np.nan, '1999-01-30', '2000-11-30'],
    'Weight (kg)': [70, np.nan, 80, 40, 90],
    'Health Insurance': [True, False, True, True, np.nan]
}

patience_infodf = pd.DataFrame(patients_info)

In [ ]:
print(patience_infodf.dtypes)
print(patience_infodf)

Last Name            object
First Name           object
Birth Date           object
Weight (kg)         float64
Health Insurance     object
dtype: object
    Last Name First Name  Birth Date  Weight (kg) Health Insurance
0      Cookie    Cadence  1977-07-07         70.0             True
1       Mango      Candy  1990-08-21          NaN            False
2     Sherbet     Pookie         NaN         80.0             True
3     Shining       Jack  1999-01-30         40.0             True
4  Hollyberry       Meow  2000-11-30         90.0              NaN


# III. Reading from file

12. Завантажити файл missile_attacks_daily.csv з [Kaggle](https://www.kaggle.com/datasets/piterfm/massive-missile-attacks-on-ukraine?select=missile_attacks_daily.csv). Ознайомитись з описом датасету. Додати файл до папки поточної сесії colab notebook на панелі ліворуч або на гугл диск (локально до проєкту, якщо працюєте у пайчармі). Зчитати датафрейм з файлу та вивести датафрейм на екран. Зверніть увагу на те, щоб всі колонки відповідали наведеним у описі типам даних.

In [5]:
attacks_data = pd.read_csv('missile_attacks_daily.csv')

print(attacks_data.dtypes)

time_start               object
time_end                 object
model                    object
launch_place             object
target                   object
carrier                  object
launched                float64
destroyed                 int64
not_reach_goal          float64
cross_border_belarus    float64
back_russia             float64
still_attacking         float64
destroyed_details        object
launched_details         object
launch_place_details     object
source                   object
dtype: object


# IV. Data understanding

13. Виведіть на екран перші 10 рядків зчитаного датафрейму.

In [4]:
print(attacks_data.head(10))

         time_start          time_end             model  \
0  2025-01-25 19:00  2025-01-26 09:00    Shahed-136/131   
1  2025-01-24 19:00  2025-01-25 09:30    Shahed-136/131   
2  2025-01-24 18:30  2025-01-24 19:30         X-59/X-69   
3  2025-01-23 21:00  2025-01-24 09:00    Shahed-136/131   
4  2025-01-22 19:00  2025-01-23 09:30    Shahed-136/131   
5  2025-01-23 04:00  2025-01-23 04:30  Iskander-M/KN-23   
6  2025-01-21 19:00  2025-01-22 10:00    Shahed-136/131   
7  2025-01-20 20:00  2025-01-21 09:00    Shahed-136/131   
8  2025-01-20 20:00  2025-01-21 09:00        Iskander-M   
9  2025-01-19 19:00  2025-01-20 08:00    Shahed-136/131   

                                        launch_place        target carrier  \
0  Primorsko-Akhtarsk and Kursk oblast and Bryans...       Ukraine     NaN   
1  Primorsko-Akhtarsk and Kursk oblast and Bryans...       Ukraine     NaN   
2                                          Black Sea  Odesa oblast     NaN   
3  Primorsko-Akhtarsk and Kursk oblast

14. Отримайте інформацію про форму датафрейму.

In [ ]:
print(attacks_data.shape)

(2056, 16)


15. Отримайте дані про кожну колонку з її назвою, типом даних та кількістю непустих значень.

In [ ]:
attacks_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2056 entries, 0 to 2055
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   time_start            2056 non-null   object 
 1   time_end              2056 non-null   object 
 2   model                 2056 non-null   object 
 3   launch_place          1034 non-null   object 
 4   target                2028 non-null   object 
 5   carrier               225 non-null    object 
 6   launched              2053 non-null   float64
 7   destroyed             2056 non-null   int64  
 8   not_reach_goal        164 non-null    float64
 9   cross_border_belarus  32 non-null     float64
 10  back_russia           35 non-null     float64
 11  still_attacking       30 non-null     float64
 12  destroyed_details     301 non-null    object 
 13  launched_details      4 non-null      object 
 14  launch_place_details  4 non-null      object 
 15  source               

16. Отримайте статистику для колонок числового типу.

In [ ]:
attacks_data.describe()

launched    destroyed  not_reach_goal  cross_border_belarus  \
count  2053.000000  2056.000000      164.000000             32.000000   
mean     10.872382     7.647860       21.231707              2.312500   
std      21.605295    14.506397       19.812804              2.054696   
min       1.000000     0.000000        1.000000              1.000000   
25%       1.000000     1.000000        4.000000              1.000000   
50%       2.000000     1.000000       16.000000              2.000000   
75%       9.000000     7.000000       30.000000              2.000000   
max     193.000000    99.000000      105.000000             10.000000   

       back_russia  still_attacking  
count    35.000000        30.000000  
mean      2.000000         2.766667  
std       1.137593         2.848876  
min       1.000000         1.000000  
25%       1.000000         1.000000  
50%       2.000000         1.000000  
75%       3.000000         3.750000  
max       5.000000        10.000000

17. Отримати кількість унікальних значень у кожній колонці. Вивести на екран значення та кількість цих значень у колонці, яка має найменше унікальних значень.

In [ ]:
unique_values = attacks_data.nunique()
print(unique_values)

time_start               874
time_end                1025
model                     61
launch_place             136
target                    97
carrier                   58
launched                 117
destroyed                 83
not_reach_goal            56
cross_border_belarus       7
back_russia                5
still_attacking            8
destroyed_details        297
launched_details           4
launch_place_details       4
source                   968
dtype: int64


In [ ]:
min_unique_column = unique_values.idxmin()

print(attacks_data[min_unique_column].value_counts())

launched_details
{'X-101/X-555': 50, 'Kalibr': 12}                     1
{'X-101/X-555': 55, 'Kalibr': 24, 'Iskander-K': 7}    1
{'X-101/X-555': 57, 'Kalibr': 28}                     1
{'X-101/X-555': 77, 'Kalibr': 28, 'X-59/X-69': 10}    1
Name: count, dtype: int64


18. На основі отриманої інформації про датасет вище, опишіть по пунктах, що б ви виправили або на що б звернули увагу на кроці очищення даних і чому?

**місце для текстової відповіді**

1.   Перевірити чи sources досі актуальні і існують, чи за посиланнями можна перейти
2.   Перевірка на нелогічні значення, типу від'ємних
3.   Є багато пустих значень, можливо заповнити плейсхолдерами
4.   Очистити дублікати якщо є




...


# V. Data Cleaning

19. Перетворити колонку launched у цілочислену (Int64). Перевірити, чи все пройшло, як очікувалось.

In [ ]:
attacks_data['launched'] = attacks_data['launched'].astype('Int64')

In [ ]:
print(attacks_data['launched'].dtype)

Int64


20. Видалити колонки, де кількість непустих значень становить менше 30% від загальної кількості записів.

In [ ]:
non_null_percents = attacks_data.notnull().mean()
attacks_data_cleaned = attacks_data.loc[:, non_null_percents >= 0.3]

print(attacks_data_cleaned)

            time_start          time_end               model  \
0     2025-01-25 19:00  2025-01-26 09:00      Shahed-136/131   
1     2025-01-24 19:00  2025-01-25 09:30      Shahed-136/131   
2     2025-01-24 18:30  2025-01-24 19:30           X-59/X-69   
3     2025-01-23 21:00  2025-01-24 09:00      Shahed-136/131   
4     2025-01-22 19:00  2025-01-23 09:30      Shahed-136/131   
...                ...               ...                 ...   
2051  2022-10-01 12:00  2022-10-01 12:00            Orlan-10   
2052        2022-09-30        2022-09-30            Orlan-10   
2053  2022-09-29 23:00  2022-09-29 23:59      Shahed-136/131   
2054        2022-09-28        2022-09-28                X-59   
2055        2022-09-28        2022-09-28  Reconnaissance UAV   

                                           launch_place  \
0     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
1     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
2                                             Black Se

21. Очистити датафрейм від дублікатів (дублюючихся рядків) та записів, де атака продовжувалась відʼємну або нульову кількість часу (що фізично неможливо). Не забудьте оновити індекс. Скільки рядків було видалено? Як ви це перевірили у коді?

In [19]:
attacks_data_cleaned = attacks_data.drop_duplicates()
attacks_data_cleaned = attacks_data_cleaned[attacks_data_cleaned['time_end'] > attacks_data_cleaned['time_start']]
attacks_data_cleaned.reset_index(drop=True, inplace=True)

rows_removed = len(attacks_data) - len(attacks_data_cleaned)

print(f"Removed {rows_removed} rows")
print(attacks_data_cleaned)

Removed 841 rows
            time_start          time_end           model  \
0     2025-01-25 19:00  2025-01-26 09:00  Shahed-136/131   
1     2025-01-24 19:00  2025-01-25 09:30  Shahed-136/131   
2     2025-01-24 18:30  2025-01-24 19:30       X-59/X-69   
3     2025-01-23 21:00  2025-01-24 09:00  Shahed-136/131   
4     2025-01-22 19:00  2025-01-23 09:30  Shahed-136/131   
...                ...               ...             ...   
1210  2022-10-06 09:00  2022-10-06 11:30       Mohajer-6   
1211        2022-10-06  2022-10-06 02:00  Shahed-136/131   
1212        2022-10-05  2022-10-05 02:00  Shahed-136/131   
1213  2022-10-02 00:00  2022-10-02 02:00  Shahed-136/131   
1214  2022-09-29 23:00  2022-09-29 23:59  Shahed-136/131   

                                           launch_place  \
0     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
1     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
2                                             Black Sea   
3     Primorsko-Akhtarsk a

Було видалено 841 рядок і я це перевірила порівнявши attack_data до та після очищення


22. Виведіть на екран всі рядки, де є хоч одне пропущене значення.

In [5]:
print(attacks_data[attacks_data.isnull().any(axis=1)])

            time_start          time_end               model  \
0     2025-01-25 19:00  2025-01-26 09:00      Shahed-136/131   
1     2025-01-24 19:00  2025-01-25 09:30      Shahed-136/131   
2     2025-01-24 18:30  2025-01-24 19:30           X-59/X-69   
3     2025-01-23 21:00  2025-01-24 09:00      Shahed-136/131   
4     2025-01-22 19:00  2025-01-23 09:30      Shahed-136/131   
...                ...               ...                 ...   
2051  2022-10-01 12:00  2022-10-01 12:00            Orlan-10   
2052        2022-09-30        2022-09-30            Orlan-10   
2053  2022-09-29 23:00  2022-09-29 23:59      Shahed-136/131   
2054        2022-09-28        2022-09-28                X-59   
2055        2022-09-28        2022-09-28  Reconnaissance UAV   

                                           launch_place  \
0     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
1     Primorsko-Akhtarsk and Kursk oblast and Bryans...   
2                                             Black Se